types/src/v0/header.rs

```diff
@@ -1,20 +1,166 @@
+                    EitherOrVersion::Left(cfg) => Ok(Header::V1(
+                        v0_1::Header::deserialize_with_chain_config(cfg.into(), seq)?,
+                    )),
+                    EitherOrVersion::Right(commit) => Ok(Header::V1(
+                        v0_1::Header::deserialize_with_chain_config(commit.into(), seq)?,
+                    )),
+                    // For all versions, the first "field" is not actually part of the `Header` struct (you can think of it as the first field of the virtual
+                    // `VersionedHeader` struct, with the `Header` being the second field. We just delegate directly to the derived deserialization
+                    // impl for the appropriate version.
+                    EitherOrVersion::Version(Version { major: 0, minor: 2 }) => Ok(Header::V2(
+                        seq.next_element()?
+                            .ok_or_else(|| de::Error::missing_field("fields"))?,
+                    )),
+                    EitherOrVersion::Version(Version { major: 0, minor: 3 }) => Ok(Header::V3(
+                        seq.next_element()?
+                            .ok_or_else(|| de::Error::missing_field("fields"))?,
+                    )),
+                    _ => Err(serde::de::Error::custom("invalid version")),
+                }
+            }
+
+            fn visit_map<V>(self, mut map: V) -> Result<Header, V::Error>
+            where
+                V: MapAccess<'de>,
+            {
+                // retrieving the first entry from the map, which could either be a
+                // `chain_config` (for v0.1 headers) or a `version` (for `VersionedHeader` structs).
+                let Some((_, chain_config_or_version)) =
+                    map.next_entry::<String, ResolvableChainConfigOrVersion>()?
+                else {
+                    return Err(de::Error::missing_field("version"));
+                };
+
+                // Match to determine the appropriate header version.
+                match chain_config_or_version.chain_config {
+                    EitherOrVersion::Left(cfg) => Ok(Header::V1(
+                        v0_1::Header::deserialize_with_chain_config_map(cfg.into(), map)?,
+                    )),
+                    EitherOrVersion::Right(commit) => Ok(Header::V1(
+                        v0_1::Header::deserialize_with_chain_config_map(commit.into(), map)?,
+                    )),
+                    EitherOrVersion::Version(Version { major: 0, minor: 2 }) => {
+                        Ok(Header::V2(map.next_entry::<String, _>()?.unwrap().1))
+                    }
+                    EitherOrVersion::Version(Version { major: 0, minor: 3 }) => {
+                        Ok(Header::V3(map.next_entry::<String, _>()?.unwrap().1))
+                    }
+                    _ => Err(serde::de::Error::custom("invalid version")),
+                }
+            }
+        }
+
+        const FIELDS: &[&str] = &["version", "fields"];
```

Comment:
So this would be all the fields for the v1 header? Is there  a better way to do this

Reply:
actually its an enum so this is wrong